### PR TITLE
Added eth namespace apis Call, EstimateGas

### PR DESCRIPTION
**Pull request by aeharvlee**
## Proposed changes

This PR implemented Eth namespace APIs `Call` and `EstimateGas`.

**Ethereum Source Code links**
* [Call](https://github.com/ethereum/go-ethereum/blob/eae3b1946a276ac099e0018fc792d9e8c3bfda6d/internal/ethapi/api.go#L968-L984)
* [EstimateGas](https://github.com/ethereum/go-ethereum/blob/eae3b1946a276ac099e0018fc792d9e8c3bfda6d/internal/ethapi/api.go#L1106-L1114)

### 1. State Override feature
* Ethereum introduced State Override option at `Call` api.
  * Now user can override the state ephemerally prior to executing the call. 
  * Related docs with detailed description: https://geth.ethereum.org/docs/rpc/ns-eth 
  * Related PR: https://github.com/ethereum/go-ethereum/pull/22245

### 2. Differences from Klaytn
* Ethereum's Call and EstimateGas basically checks if caller (= the from address) has enough balance to proceeds call if user specified `gas` and `gasPrice` field.
  * Checks balance of caller is higher than `gas * gasPrice`.

For this, I moved logic adding balance of caller from GetEVM to out of it because both `klay_call` and `eth_call` are now using same API but works differently.

Klaytn add gasFee by default before proceeding `call` and `estimateGas`.

### 3. Differences from Ethereum implementation

* **Removed ExecutionResult struct**
  * There is no easy way to classify whether the error returned from TransitionDb in Klaytn is vmError or consensusError or just simple error can be detected before executing EVM (e.g. user specify gas limit which is lower than intrinsicGas).
  * Ethereum uses ExecutionResult because they can specify that errors in their codebase, but this is not adaptable to current Klaytn codebase.
We can apply this changes maybe later.

* **AccessList is not supported yet**
  * AccessList is not supported yet in Klaytn, so it does not process with it even though user passes AccessList as parameter.

## How to test this PR

First, please clone my [origin repo](https://github.com/aeharvlee/klaytn/tree/sendtxargs-apis-0).

**1. Build ken with source codes related with this PR.**
* Run `make ken`

**2. Run ken**
* In my case, I tested with CN 1 - PN 1 - EN 1 in my local.
* Please note that you should enable `eth` namepsace to RPC_API and WS_API in kend.conf

### Prepare the test
<details>
  <summary>Click to expand</summary>

`Storage.bin-runtime`: `6080604052600436106049576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680632e64cec114604e5780636057361d146076575b600080fd5b348015605957600080fd5b50606060a0565b6040518082815260200191505060405180910390f35b348015608157600080fd5b50609e6004803603810190808035906020019092919050505060a9565b005b60008054905090565b80600081905550505600a165627a7a723058207783dba41884f73679e167576362b7277f88458815141651f48ca38c25b498f80029`

I used `$ docker run -v /Users/denver.lee/playground/solidity/storage:/sol ethereum/solc:0.4.24 --asm --abi --bin --bin-runtime --metadata --overwrite -o /sol/output /sol/storage.sol` to compile below Storage contract code.

```solidity
pragma solidity ^0.4.24;

/**
 * @title Storage
 * @dev Store & retrieve value in a variable
 */
contract Storage {

    uint256 number;

    /**
     * @dev Store value in variable
     * @param num value to store
     */
    function store(uint256 num) public {
        number = num;
    }

    /**
     * @dev Return value 
     * @return value of 'number'
     */
    function retrieve() public view returns (uint256){
        return number;
    }
}
```

#### ABI Set to test function call of KIP-7
```node
> totalSupplyAbi
{ name: 'totalSupply', type: 'function', inputs: [] }
> caver.abi.encodeFunctionCall(totalSupplyAbi, [])
'0x18160ddd' 

> balanceOfAbi
{
  constant: true,
  inputs: [ { internalType: 'address', name: 'account', type: 'address' } ],
  name: 'balanceOf',
  outputs: [ { internalType: 'uint256', name: '', type: 'uint256' } ],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}
> caver.abi.encodeFunctionCall(balanceOfAbi, ["0xca7a99380131e6c76cfa622396347107aeedca2d"])
'0x70a08231000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d'

> retrieveAbi
{ name: 'retrieve', type: 'function', inputs: [] }
> caver.abi.encodeFunctionCall(retrieveAbiObj, [])
'0x2e64cec1'
```

</details>

### Test Call
<details>
  <summary>Click to expand</summary>

**I tested this api in the following order.**
1. Deploy KIP-7 contract.
2. Call KIP-7 contract functions `totalSupply` and `balanceOf` to check whether call is working or not.
3. State Override already deployed KIP-7 contract account by replacing it's code with runtime bytecode of `Storage.sol` and call `retrieve` function of `Storage.sol`.

Here is the my test scenario info below.
* KIP-7 Contract address: `0xbE3892d33620bE5aca8c75D39e7401871194d290`
* Deployer address: `0xca7a99380131e6c76cfa622396347107aeedca2d`

**Call totalSupply()**
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_call", "params": [{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0xbE3892d33620bE5aca8c75D39e7401871194d290", "data": "0x18160ddd"}, "latest"], "id": 1}' http://localhost:8551

{"jsonrpc":"2.0","id":1,"result":"0x0000000000000000000000000000000000000000000000000de0b6b3a7640000"}
```

**Call balanceOf("0xca7a99380131e6c76cfa622396347107aeedca2d")**
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_call", "params": [{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0xbE3892d33620bE5aca8c75D39e7401871194d290", "data": "0x70a08231000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"}, "latest"], "id": 1}' http://localhost:8551

{"jsonrpc":"2.0","id":1,"result":"0x0000000000000000000000000000000000000000000000000de0b6b3a763ffff"}
```

**State Override with Storage.sol and Call retrieve**
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_call", "params": [{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0xbE3892d33620bE5aca8c75D39e7401871194d290", "data": "0x2e64cec1"}, "latest", {"0xbE3892d33620bE5aca8c75D39e7401871194d290": {"code":"0x6080604052600436106049576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680632e64cec114604e5780636057361d146076575b600080fd5b348015605957600080fd5b50606060a0565b6040518082815260200191505060405180910390f35b348015608157600080fd5b50609e6004803603810190808035906020019092919050505060a9565b005b60008054905090565b80600081905550505600a165627a7a723058207783dba41884f73679e167576362b7277f88458815141651f48ca38c25b498f80029"}}], "id": 1}' http://localhost:8551

{"jsonrpc":"2.0","id":1,"result":"0x0000000000000000000000000000000000000000000000000000000000000000"}
```

**Call with gas and gasPrice but balance is not enough**
Balance of caller `0xe2df3753d74bebd8c4ccdbd5320b36e5ef8ca08d`: `0`.
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_call", "params": [{"from": "0xe2df3753d74bebd8c4ccdbd5320b36e5ef8ca08d", "to": "0xbE3892d33620bE5aca8c75D39e7401871194d290", "data": "0x70a08231000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d", "gas": "0x99999", "gasPrice": "0x1"}, "latest"], "id": 1}' http://localhost:8551
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"err: insufficient balance of the fee payer to pay for gas (supplied gas 629145)"}}
```

**Call with gas and gasPrice but balance is not enough -> Using State Override to cover this**
But if we use State Override we can use it. Let's override state with balance `0x99999`.
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_call", "params": [{"from": "0xe2df3753d74bebd8c4ccdbd5320b36e5ef8ca08d", "to": "0xbE3892d33620bE5aca8c75D39e7401871194d290", "data": "0x70a08231000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d", "gas": "0x99999", "gasPrice": "0x1"}, "latest", {"0xe2df3753d74bebd8c4ccdbd5320b36e5ef8ca08d": {"balance": "0x99999"}}], "id": 1}' http://localhost:8551

{"jsonrpc":"2.0","id":1,"result":"0x0000000000000000000000000000000000000000000000000de0b6b3a763ffff"}
```

</details>

### Test EstimateGas
<details>
  <summary>Click to expand</summary>

**Estimate Gas with gas and gasPrice. Caller balance is enough to call it.**
Balance of caller `0xca7a99380131e6c76cfa622396347107aeedca2d`: `9.99999999999999999999999999999999097166e+49`
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "eth_estimateGas", "params": [{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0x87ac99835e67168d4f9a40580f8f5c33550ba88b", "gas": "0x100000", "gasPrice": "0x5d21dba00", "value": "0x0", "data": "0x8ada066e"}], "id": 1}' http://localhost:8551

{"jsonrpc":"2.0","id":1,"result":"0x5398"}
```

**Estimate Gas with gas and gasPrice but Caller balance is not enough.**
Balance of caller `0xe2df3753d74bebd8c4ccdbd5320b36e5ef8ca08d`: `0`
```shell
➜  corecell curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "eth_estimateGas", "params": [{"from": "0xe2df3753d74bebd8c4ccdbd5320b36e5ef8ca08d", "to": "0x87ac99835e67168d4f9a40580f8f5c33550ba88b", "gas": "0x100000", "gasPrice": "0x5d21dba00", "value": "0x0", "data": "0x8ada066e"}], "id": 1}' http://localhost:8551

{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"insufficient funds for transfer"}}
```

</details>

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: api/api_ethereum.go [hunk 1]
```diff
@@ -791,6 +843,204 @@ type EthTransactionArgs struct {
...
+	// After london, default to 1559 uncles gasPrice is set
+	head := b.CurrentBlock().Header()
+	// TODO-Klaytn: Klaytn is using fixed BaseFee(0) as now but
+	// if we apply dynamic BaseFee, we should add calculated BaseFee instead of using params.BaseFee.
+	fixedBaseFee := new(big.Int).SetUint64(params.BaseFee)
+
+	// If user specifies both maxPriorityfee and maxFee, then we do not
+	// need to consult the chain for defaults. It's definitely a London tx.
+	if args.MaxPriorityFeePerGas == nil || args.MaxFeePerGas == nil {
+		if b.ChainConfig().IsLondon(head.Number) && args.GasPrice == nil {
```
Comment (jimni1222) [inline api/api_ethereum.go:879]: Do we need to check `args.GasPrice == nil` in here?

File: api/api_ethereum.go [hunk 2]
```diff
@@ -791,6 +843,204 @@ type EthTransactionArgs struct {
...
+					// TODO-Klaytn: Klaytn is using fixed BaseFee(0) as now but
+					// if we apply dynamic BaseFee, we should add calculated BaseFee instead of params.BaseFee.
+					price.Add(price, new(big.Int).SetUint64(params.BaseFee))
+				}
+			}
+		}
+	} else {
+		// Both maxPriorityFee and maxFee set by caller. Sanity-check their internal relation
+		if args.MaxFeePerGas.ToInt().Cmp(args.MaxPriorityFeePerGas.ToInt()) < 0 {
+			return errors.New("maxFeePerGas or maxPriorityFeePerGas specified but london is not active yet")
```
Comment (jimni1222) [inline api/api_ethereum.go:924]: Is this correct error msg?

Comment (aeharvlee) [reply api/api_ethereum.go]: @jimni1222 
Thanks jamie. I fixed it [here](https://github.com/klaytn/klaytn/pull/1134/commits/67075b02705c246724248c935b0dece81dabbf91)

File: api/api_ethereum.go [hunk 3]
```diff
@@ -917,3 +1166,164 @@ func (api *EthereumAPI) rpcMarshalHeader(head *types.Header) (map[string]interfa
...
+	if timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+	// Make sure the context is cancelled when the call has completed
+	// this makes sure resources are cleaned up.
+	defer cancel()
+
+	// TODO-Klaytn: Klaytn is using fixed baseFee as now.
```
Comment (jimni1222) [inline api/api_ethereum.go:1193]: Is this correct todo comment? I don't get it what we need to do from this.

File: api/api_ethereum.go [hunk 4]
```diff
@@ -917,3 +1166,164 @@ func (api *EthereumAPI) rpcMarshalHeader(head *types.Header) (map[string]interfa
...
+	fixedBaseFee := new(big.Int).SetUint64(params.BaseFee)
+	intrinsicGas, err := types.IntrinsicGas(args.data(), args.To == nil, b.ChainConfig().Rules(header.Number))
+	if err != nil {
+		return nil, 0, err
+	}
+	msg, err := args.ToMessage(globalGasCap, fixedBaseFee, intrinsicGas)
+	if err != nil {
+		return nil, 0, err
+	}
+	evm, vmError, err := b.GetEVM(ctx, msg, st, header, vm.Config{})
```
Comment (jimni1222) [inline api/api_ethereum.go:1203]: Um.. This is just question.
There is `NoBaseFee` in vm.Config in Ethereum code.
Don't we need that?

File: api/api_ethereum.go [hunk 5]
```diff
@@ -765,6 +816,204 @@ type EthTransactionArgs struct {
...
+	addr := args.from()
+
+	// Set default gas & gas price if none were set
+	gas := globalGasCap
+	if gas == 0 {
+		gas = uint64(math.MaxUint64 / 2)
+	}
+	if args.Gas != nil {
+		gas = uint64(*args.Gas)
+	}
```
Comment (jimni1222) [inline api/api_ethereum.go:963]: Maybe change the sequence those two better?

File: api/api_ethereum.go (hunk 3 above)
Comment (aeharvlee) [reply api/api_ethereum.go:1193]: Oh, I will add some context about it.
What I wanted to say is: "Klaytn is using fixed baseFee as now. If we change this fixed baseFee as dynamic baseFee, we should update this logic too."

File: api/api_ethereum.go [hunk 6]
```diff
@@ -917,3 +1166,164 @@ func (api *EthereumAPI) rpcMarshalHeader(head *types.Header) (map[string]interfa
...
+		_, _, err := EthDoCall(ctx, b, args, rpc.NewBlockNumberOrHashWithNumber(rpc.LatestBlockNumber), nil, 0, gasCap)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	// Execute the binary search and hone in on an executable gas limit
+	for lo+1 < hi {
+		mid := (hi + lo) / 2
+		isExecutable, _ := executable(mid)
```
Comment (jimni1222) [inline api/api_ethereum.go:1310]: We don't need to check err here?
https://github.com/ethereum/go-ethereum/blob/eae3b1946a276ac099e0018fc792d9e8c3bfda6d/internal/ethapi/api.go#L1077

File: api/api_public_blockchain.go [hunk 7]
```diff
@@ -295,6 +295,8 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.Blo
 	// this makes sure resources are cleaned up.
 	defer cancel()
 
+	// Add gas fee to sender for estimating gasLimit/computing cost or calling a function by insufficient balance sender.
+	state.AddBalance(msg.ValidatedSender(), new(big.Int).Mul(new(big.Int).SetUint64(msg.Gas()), msg.GasPrice()))
```
Comment (jimni1222) [inline api/api_public_blockchain.go:299]: Before adding this, what happen when sender does not have enough balance?

File: blockchain/state/statedb.go [hunk 8]
```diff
@@ -471,6 +471,13 @@ func (self *StateDB) SetState(addr common.Address, key, value common.Hash) {
 	}
 }
 
+func (self *StateDB) SetStorage(addr common.Address, storage map[common.Hash]common.Hash) {
```
Comment (jimni1222) [inline blockchain/state/statedb.go:474]: Can you add some comments for this?

File: api/api_ethereum.go (hunk 4 above)
Comment (aeharvlee) [reply api/api_ethereum.go:1203]: That option is used in [preCheck](https://github.com/ethereum/go-ethereum/blob/fb3a6528cfa49f623570575c4fe9e8a716cfcdf7/core/state_transition.go#L238-L256) to determine whether or not to do a sanity check about London related values.

I'd like to share the scope of my PR.
* I tried not to change data structures related to the StateTransition part as much as possible. 
* It because that is core part of Klaytn and need some lots of change related with it.
* For this, I used the GasPrice field only. 
  * Even if user specify MaxPriorityPerGas and MaxFeePerGas, GasPrice field is calculated and used based on that  values.

File: api/api_ethereum.go (hunk 3 above)
Comment (aeharvlee) [reply api/api_ethereum.go]: Added some context [in here](https://github.com/klaytn/klaytn/pull/1134/commits/15cb596fffac22f51e4841d7f16f9dbe981d5b54)

File: api/api_ethereum.go (hunk 5 above)
Comment (aeharvlee) [reply api/api_ethereum.go:963]: Looks great :) I'll update this PR soon.

Comment (aeharvlee) [reply api/api_ethereum.go:963]: @jimni1222 
Oh.. changing sequence looks not great. sorry for reversal.
Let's consider the situation user give gas as `0`.

If we switch that sequence, 
gas = `uint64(math.MaxUint64 / 2)` applied and it is not what caller intended.

File: api/api_public_blockchain.go (hunk 7 above)
Comment (aeharvlee) [reply api/api_public_blockchain.go:299]: @jimni1222 
I moved this line from [here](https://github.com/klaytn/klaytn/pull/1134/files#diff-c7deb6fb16afa13f94bea51c3e7d8d9da0a069beb20036f94a6bae90442a499dL209-L210).

It because now EthDoCall (used by eth_call) and DoCall (used by klay_call) both using GetEVM and works differently.
* EthDoCall requires caller have enough gas to call.
* DoCall requires no gas to call.

If we keep that line in GetEVM, it is hard to support that differences.

**Comment by jimni1222** [comment]
Please resolve conflicts

File: api/api_ethereum.go (hunk 6 above)
Comment (aeharvlee) [reply api/api_ethereum.go]: To handle error just like Ethereum logic, I updated some codes in [this commit](https://github.com/klaytn/klaytn/pull/1134/commits/0c0f6c8792c4b9a8b4ee62a5e3538dad36870ff3).

* Now we can specify whether the error is occurred by EVM execution or not.

After that commit,

**KEN: Test with revert situation(Call function required owner from address which is not owner)**
```js
> eth.estimateGas({from: "0x3753f5bf8cCA929EeeB4B74F200dfc6375ad5444", to: "0xbe3892d33620be5aca8c75d39e7401871194d290", data: "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"})


Error: execution reverted: Caller is not owner
    at web3.js:3278:20
    at web3.js:6805:15
    at web3.js:5216:36
    at <anonymous>:1:1
```

**Geth: Test with revert situation((Call function required owner from address which is not owner)**
```js
> eth.estimateGas({from: "0xca7a99380131e6c76cfa622396347107aeedca2d", to: "0x27a110C35BA478034d7F874C89787E0FE587f65F", data: "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"})
Error: execution reverted: Caller is not owner
	at web3.js:6357:37(47)
	at web3.js:5091:62(37)
	at <eval>:1:16(10)
```


FYI. In klay namespace, it cannot specify the error is occurred by VM or not.
> klay.estimateGas({from: "0x3753f5bf8cCA929EeeB4B74F200dfc6375ad5444", to: "0xbe3892d33620be5aca8c75d39e7401871194d290", data: "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"})


Error: gas required exceeds allowance or always failing transaction
    at web3.js:3278:20
    at web3.js:6805:15
    at web3.js:5216:36
    at <anonymous>:1:1

Comment (jimni1222) [reply api/api_ethereum.go]: Ohhh... looks good! This might be good to applied in klay in the future. Thank you

File: blockchain/state/statedb.go (hunk 8 above)
Comment (aeharvlee) [reply blockchain/state/statedb.go:476]: Added comment in [this commit](https://github.com/klaytn/klaytn/pull/1134/commits/55b832f760f17c68c4f595c7d6feffbd7d7437ac).

File: api/api_ethereum.go [hunk 9]
```diff
@@ -1429,13 +1445,62 @@ func EthDoEstimateGas(ctx context.Context, b Backend, args EthTransactionArgs, b
...
+	}
+	return &revertError{
+		error:  err,
+		reason: hexutil.Encode(data),
+	}
+}
+
+// revertError is an API error that encompassas an EVM revertal with JSON error
+// code and a binary data blob.
+type revertError struct {
```
Comment (jimni1222) [inline api/api_ethereum.go:1493]: Don't we need to implement `Error() string` for this?

File: api/api_ethereum.go [hunk 10]
```diff
@@ -1029,3 +1282,226 @@ func (api *EthereumAPI) rpcMarshalBlock(block *types.Block, inclTx, fullTx bool)
...
+			}
+			// Otherwise, the specified gas cap is too low
+			return 0, fmt.Errorf("gas required exceeds allowance (%d)", cap)
+		}
+	}
+	return hexutil.Uint64(hi), nil
+}
+
+// isVMError returns true if given error is occurred during EVM execution.
+func isVMError(err error) bool {
```
Comment (jimni1222) [inline api/api_ethereum.go:1468]: Do you think we can move this to `vm` package?
Then we don't need vmErrors variable and also easier to maintain for adding new vm errors in the future.

File: api/api_ethereum.go [hunk 11]
```diff
@@ -1029,3 +1282,226 @@ func (api *EthereumAPI) rpcMarshalBlock(block *types.Block, inclTx, fullTx bool)
...
+	}
+	// Reject the transaction as invalid if it still fails at the highest allowance
+	if hi == cap {
+		isExecutable, ret, vmErr, err := executable(hi)
+		if err != nil {
+			return 0, err
+		}
+		if !isExecutable {
+			if ret != nil {
+				if vmErr == vm.ErrExecutionReverted || strings.Contains(vmErr.Error(), vm.ErrExecutionReverted.Error()) {
```
Comment (jimni1222) [inline api/api_ethereum.go:1455]: This is just question, can you explain why did you add `strings.Contains(vmErr.Error(), vm.ErrExecutionReverted.Error())` to check `vmErr` is `vm.ErrExecutionReverted`?
Is there any logic to wrap vm error in `EthDoCall`?

Comment (jimni1222) [reply api/api_ethereum.go:1455]: Please don't mind this comment

File: api/api_ethereum.go (hunk 9 above)
Comment (aeharvlee) [reply api/api_ethereum.go:1493]: @jimni1222 
I'll update this PR soon. I should have checked via http call, not only via console. 
Thanks a lot.

File: api/api_ethereum.go [hunk 12]
```diff
@@ -1029,3 +1282,226 @@ func (api *EthereumAPI) rpcMarshalBlock(block *types.Block, inclTx, fullTx bool)
...
+}
+
+// ErrorCode returns the JSON error code for a revertal.
+// See: https://github.com/ethereum/wiki/wiki/JSON-RPC-Error-Codes-Improvement-Proposal
+func (e *revertError) ErrorCode() int {
+	return 3
+}
+
+// ErrorData returns the hex encoded revert reason.
+func (e *revertError) ErrorData() interface{} {
```
Comment (jimni1222) [inline api/api_ethereum.go:1505]: I think we can remove this function

File: api/api_ethereum.go (hunk 9 above)
Comment (aeharvlee) [reply api/api_ethereum.go:1493]: I applied that in this [commit](https://github.com/klaytn/klaytn/pull/1134/commits/2769453baa2c0b2d2d73e07602004bc5efdfdd63).

After this commit, we can get exactly same return format with Geth.
**KEN:**
```shell
$ curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_estimateGas", "params": [{"from": "0x3753f5bf8cCA929EeeB4B74F200dfc6375ad5444", "to": "0xbe3892d33620be5aca8c75d39e7401871194d290", "data": "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"}], "id": 1}' http://localhost:8551
{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"execution reverted: Caller is not owner","data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343616c6c6572206973206e6f74206f776e657200000000000000000000000000"}}
```

**Geth:**
```shell
$ curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "method": "eth_estimateGas", "params": [{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0x27a110C35BA478034d7F874C89787E0FE587f65F", "data": "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"}], "id": 1}' http://localhost:8545
{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"execution reverted: Caller is not owner","data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343616c6c6572206973206e6f74206f776e657200000000000000000000000000
```

Related Ethereum's PR is https://github.com/ethereum/go-ethereum/pull/22677

File: api/api_ethereum.go (hunk 12 above)
Comment (aeharvlee) [reply api/api_ethereum.go:1505]: Please check [this comment](https://github.com/klaytn/klaytn/pull/1134#discussion_r803563714) above :)

I updated PR related with that.

File: api/api_ethereum.go (hunk 9 above)
Comment (aidan-kwon) [reply api/api_ethereum.go:1504]: @aeharvlee I think we need to implement in klay namspace later. Could you make a task issue for v1.9.0?

File: api/api_ethereum.go [hunk 13]
```diff
@@ -1029,3 +1285,234 @@ func (api *EthereumAPI) rpcMarshalBlock(block *types.Block, inclTx, fullTx bool)
...
+		}
+		balance := state.GetBalance(*args.From) // from can't be nil
+		available := new(big.Int).Set(balance)
+		if args.Value != nil {
+			if args.Value.ToInt().Cmp(available) >= 0 {
+				return 0, errors.New("insufficient funds for transfer")
+			}
+			available.Sub(available, args.Value.ToInt())
+		}
+		allowance := new(big.Int).Div(available, feeCap)
```
Comment (aidan-kwon) [inline api/api_ethereum.go:1397]: I think it need to be `allowance := new(big.Int).Sub(available, feeCap)`. How do you think? Even `feeCap` can be zero and occurs panic.

**Review by aidan-kwon — COMMENTED** [review]
I think it's good except for the comment I leaved

**Comment by aidan-kwon** [comment]
@jeongkyun-oh @yoomee1313 PTAL

File: api/api_ethereum.go (hunk 10 above)
Comment (aeharvlee) [reply api/api_ethereum.go]: Updated in [this commit](https://github.com/klaytn/klaytn/pull/1134/commits/51faecf814b6c99777904d651b51cedfe05ee352).

Please take a look :)

File: api/api_ethereum.go (hunk 13 above)
Comment (aeharvlee) [reply api/api_ethereum.go:1396]: feeCap can't be zero in this scope.
```
```

**Review by jimni1222 — APPROVED** [review]

File: api/api_ethereum.go (hunk 13 above)
Comment (aidan-kwon) [reply api/api_ethereum.go:1396]: Oh, I see. Thank you for the explanation.

**Review by aidan-kwon — APPROVED** [review]

File: api/api_ethereum.go [hunk 14]
```diff
@@ -1029,3 +1284,224 @@ func (api *EthereumAPI) rpcMarshalBlock(block *types.Block, inclTx, fullTx bool)
...
+	fixedBaseFee := new(big.Int).SetUint64(params.BaseFee)
+	intrinsicGas, err := types.IntrinsicGas(args.data(), args.To == nil, b.ChainConfig().Rules(header.Number))
+	if err != nil {
+		return nil, 0, err
+	}
+	msg, err := args.ToMessage(globalGasCap, fixedBaseFee, intrinsicGas)
+	if err != nil {
+		return nil, 0, err
+	}
+	evm, vmError, err := b.GetEVM(ctx, msg, st, header, vm.Config{})
```
Comment (sirano11) [inline api/api_ethereum.go:1321]: Would you check that need to call a `state.AddBalance()` function before calling `GetEVM()`? 
You moved the code `state.AddBalance()` from `GetEVM()`.

Comment (jimni1222) [reply api/api_ethereum.go:1321]: You can check [here](https://github.com/klaytn/klaytn/pull/1134#discussion_r802444765)

**Review by sirano11 — APPROVED** [review]
